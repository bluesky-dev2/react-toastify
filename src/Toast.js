import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { css } from 'glamor';

import ProgressBar from './ProgressBar';
import { POSITION, TYPE } from './constant';
import defaultStyle from './defaultStyle';
import {
  falseOrElement,
  falseOrDelay,
  objectValues
} from './util/propValidator';

const styles = {
  container: type =>
    css({
      position: 'relative',
      minHeight: '48px',
      marginBottom: '1rem',
      padding: '8px',
      borderRadius: '1px',
      boxShadow:
        '0 1px 10px 0 rgba(0, 0, 0, .1), 0 2px 15px 0 rgba(0, 0, 0, .05)',
      display: 'flex',
      justifyContent: 'space-between',
      maxHeight: '800px',
      overflow: 'hidden',
      fontFamily: defaultStyle.fontFamily,
      cursor: 'pointer',
      background:
        defaultStyle[`color${type.charAt(0).toUpperCase()}${type.slice(1)}`],
      ...(type === 'default' ? { color: '#aaa' } : {}),
      [`@media ${defaultStyle.mobile}`]: {
        marginBottom: 0
      }
    }),
  body: css({
    margin: 'auto 0',
    flex: 1
  })
};

class Toast extends Component {
  static propTypes = {
    closeButton: falseOrElement.isRequired,
    autoClose: falseOrDelay.isRequired,
    children: PropTypes.node.isRequired,
    closeToast: PropTypes.func.isRequired,
    position: PropTypes.oneOf(objectValues(POSITION)).isRequired,
    pauseOnHover: PropTypes.bool.isRequired,
    closeOnClick: PropTypes.bool.isRequired,
    transition: PropTypes.func.isRequired,
    isDocumentHidden: PropTypes.bool.isRequired,
    in: PropTypes.bool,
    onExited: PropTypes.func,
    hideProgressBar: PropTypes.bool,
    onOpen: PropTypes.func,
    onClose: PropTypes.func,
    type: PropTypes.oneOf(objectValues(TYPE)),
    className: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    bodyClassName: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    progressClassName: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.object
    ]),
    updateId: PropTypes.number,
    ariaLabel: PropTypes.string
  };

  static defaultProps = {
    type: TYPE.DEFAULT,
    in: true,
    hideProgressBar: false,
    onOpen: null,
    onClose: null,
    className: '',
    bodyClassName: '',
    progressClassName: '',
    updateId: null,
    role: 'alert'
  };

  state = {
    isRunning: true
  };

  componentDidMount() {
    this.props.onOpen !== null && this.props.onOpen(this.getChildrenProps());
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.isDocumentHidden !== nextProps.isDocumentHidden) {
      this.setState({
        isRunning: !nextProps.isDocumentHidden
      });
    }
  }

  componentWillUnmount() {
    this.props.onClose !== null && this.props.onClose(this.getChildrenProps());
  }

  getChildrenProps() {
    return this.props.children.props;
  }

  getToastProps() {
    const toastProps = {};

    if (this.props.autoClose !== false && this.props.pauseOnHover === true) {
      toastProps.onMouseEnter = this.pauseToast;
      toastProps.onMouseLeave = this.playToast;
    }
    typeof this.props.className === 'string' &&
      (toastProps.className = this.props.className);
    this.props.closeOnClick && (toastProps.onClick = this.props.closeToast);

    return toastProps;
  }

  pauseToast = () => {
    this.setState({ isRunning: false });
  };

  playToast = () => {
    this.setState({ isRunning: true });
  };

  render() {
    const {
      closeButton,
      children,
      autoClose,
      type,
      hideProgressBar,
      closeToast,
      transition: Transition,
      position,
      onExited,
      className,
      bodyClassName,
      progressClassName,
      updateId,
      role
    } = this.props;

    return (
      <Transition
        in={this.props.in}
        appear
        unmountOnExit
        onExited={onExited}
        position={position}
      >
        <div
          {...(typeof className !== 'string'
            ? css(styles.container(type), className)
            : styles.container(type))}
          {...this.getToastProps()}
        >
          <div
            {...this.props.in && { role: role }}
            {...(typeof bodyClassName !== 'string'
              ? css(styles.body, bodyClassName)
              : styles.body)}
            {...typeof bodyClassName === 'string' && {
              className: bodyClassName
            }}
          >
            {children}
          </div>
          {closeButton !== false && closeButton}
          {autoClose !== false && (
            <ProgressBar
              key={`pb-${updateId}`}
              delay={autoClose}
              isRunning={this.state.isRunning}
              closeToast={closeToast}
              hide={hideProgressBar}
              type={type}
              className={progressClassName}
            />
          )}
        </div>
      </Transition>
    );
  }
}

export default Toast;
