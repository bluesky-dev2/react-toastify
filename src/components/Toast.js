import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import ProgressBar from './ProgressBar';
import { POSITION, TYPE } from './../utils/constant';
import {
  falseOrElement,
  falseOrDelay,
  objectValues
} from './../utils/propValidator';


function getX(e) {
  return e.targetTouches && e.targetTouches.length >= 1
    ? e.targetTouches[0].clientX
    : e.clientX;
}

function getY(e){
  return e.targetTouches && e.targetTouches.length >= 1
    ? e.targetTouches[0].clientY
    : e.clientY;
}

class Toast extends Component {
  static propTypes = {
    closeButton: falseOrElement.isRequired,
    autoClose: falseOrDelay.isRequired,
    children: PropTypes.node.isRequired,
    closeToast: PropTypes.func.isRequired,
    position: PropTypes.oneOf(objectValues(POSITION)).isRequired,
    pauseOnHover: PropTypes.bool.isRequired,
    closeOnClick: PropTypes.bool.isRequired,
    transition: PropTypes.func.isRequired,
    isDocumentHidden: PropTypes.bool.isRequired,
    rtl: PropTypes.bool.isRequired,
    in: PropTypes.bool,
    onExited: PropTypes.func,
    hideProgressBar: PropTypes.bool,
    onOpen: PropTypes.func,
    onClose: PropTypes.func,
    type: PropTypes.oneOf(objectValues(TYPE)),
    className: PropTypes.string,
    bodyClassName: PropTypes.string,
    progressClassName: PropTypes.string,
    updateId: PropTypes.number,
    draggable: PropTypes.bool,
    draggablePercent: PropTypes.number,
    ariaLabel: PropTypes.string
  };

  static defaultProps = {
    type: TYPE.DEFAULT,
    in: true,
    hideProgressBar: false,
    onOpen: null,
    onClose: null,
    className: null,
    bodyClassName: null,
    progressClassName: null,
    updateId: null,
    draggable: true,
    draggablePercent: 0.8,
    role: 'alert'
  };

  state = {
    isRunning: true,
    preventExitTransition: false
  };

  flags = {
    canCloseOnClick: true,
    canDrag: false
  };

  position = {
    start: 0,
    x: 0,
    y: 0,
    deltaX: 0
  };

  ref = null;

  removalDistance = -1;

  componentDidMount() {
    this.props.onOpen !== null && this.props.onOpen(this.getChildrenProps());
    if (this.props.draggable) {
      document.addEventListener('mousemove', this.onDragMove);
      document.addEventListener('mouseup', this.onDragEnd);

      document.addEventListener('touchmove', this.onDragMove);
      document.addEventListener('touchend', this.onDragEnd);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.isDocumentHidden !== nextProps.isDocumentHidden) {
      this.setState({
        isRunning: !nextProps.isDocumentHidden
      });
    }
  }

  componentWillUnmount() {
    this.props.onClose !== null && this.props.onClose(this.getChildrenProps());
    if (this.props.draggable) {
      document.removeEventListener('mousemove', this.onDragMove);
      document.removeEventListener('mouseup', this.onDragEnd);

      document.removeEventListener('touchmove', this.onDragMove);
      document.removeEventListener('touchend', this.onDragEnd);
    }
  }

  pauseToast = () => {
    this.setState({ isRunning: false });
  };

  playToast = () => {
    this.setState({ isRunning: true });
  };

  getChildrenProps() {
    return this.props.children.props;
  }

  getToastProps() {
    const toastProps = {};
    const { autoClose, pauseOnHover, closeOnClick, closeToast } = this.props;

    if (autoClose && pauseOnHover) {
      toastProps.onMouseEnter = this.pauseToast;
      toastProps.onMouseLeave = this.playToast;
    }

    // prevent toast from closing when user drags the toast
    if (closeOnClick) {
      toastProps.onClick = () => this.flags.canCloseOnClick && closeToast();
    }

    return toastProps;
  }

  onDragStart = e => {
    this.flags.canCloseOnClick = true;
    this.flags.canDrag = true;
    this.ref.style.transition = '';
    this.position.start = this.position.x = getX(e.nativeEvent);
    this.removalDistance = this.ref.offsetWidth * this.props.draggablePercent;
  };

  onDragMove = e => {
    if (this.flags.canDrag) {

      if (this.state.isRunning) {
        this.pauseToast();
      }

      this.position.x = getX(e);
      this.position.deltaX = this.position.x - this.position.start;
      //prevent false positif during a toast click
      this.position.start !== this.position.x &&
        (this.flags.canCloseOnClick = false);

      this.ref.style.transform = `translateX(${this.position.deltaX}px)`;
      this.ref.style.opacity =
        1 - Math.abs(this.position.deltaX / this.removalDistance);
    }
  };

  onDragEnd = e => {
    if (this.flags.canDrag) {
      this.flags.canDrag = false;

      if (Math.abs(this.position.deltaX) > this.removalDistance) {
        this.setState(
          {
            preventExitTransition: true
          },
          this.props.closeToast
        );
        return;
      }

      this.position.y = getY(e);
      this.ref.style.transition = 'transform 0.2s, opacity 0.2s';
      this.ref.style.transform = 'translateX(0)';
      this.ref.style.opacity = 1;
    }
  };

  onDragTransitionEnd = () => {
    const { top, bottom, left, right } = this.ref.getBoundingClientRect();

    if (
      this.props.pauseOnHover &&
      this.position.x >= left &&
      this.position.x <= right &&
      this.position.y >= top &&
      this.position.y <= bottom
    ) {
      this.pauseToast();
    } else {
      this.playToast();
    }
  };

  render() {
    const {
      closeButton,
      children,
      autoClose,
      type,
      hideProgressBar,
      closeToast,
      transition: Transition,
      position,
      onExited,
      className,
      bodyClassName,
      progressClassName,
      updateId,
      role,
      rtl
    } = this.props;

    const toastClassname = cx(
      'Toastify__toast',
      `Toastify__toast--${type}`,
      className,
      {
        'Toastify__toast--rtl': rtl
      }
    );

    const bodyClassname = cx('Toastify__toast-body', bodyClassName);
    return (
      <Transition
        in={this.props.in}
        appear
        unmountOnExit
        onExited={onExited}
        position={position}
        preventExitTransition={this.state.preventExitTransition}
      >
        <div
          className={toastClassname}
          {...this.getToastProps()}
          ref={ref => (this.ref = ref)}
          onMouseDown={this.onDragStart}
          onTouchStart={this.onDragStart}
          onTransitionEnd={this.onDragTransitionEnd}
        >
          <div {...this.props.in && { role: role }} className={bodyClassname}>
            {children}
          </div>
          {closeButton !== false && closeButton}
          {autoClose !== false && (
            <ProgressBar
              key={`pb-${updateId}`}
              rtl={rtl}
              delay={autoClose}
              isRunning={this.state.isRunning}
              closeToast={closeToast}
              hide={hideProgressBar}
              type={type}
              className={progressClassName}
            />
          )}
        </div>
      </Transition>
    );
  }
}

export default Toast;
