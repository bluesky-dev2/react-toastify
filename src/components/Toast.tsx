import * as React from 'react';
import cx from 'clsx';

import { ProgressBar } from './ProgressBar';
import { ToastProps } from '../types';
import { Default, isFn } from '../utils';
import { useToast } from '../hooks';

export const Toast: React.FC<ToastProps> = props => {
  const {
    isRunning,
    preventExitTransition,
    toastRef,
    eventHandlers
  } = useToast(props);
  const {
    closeButton,
    children,
    autoClose,
    onClick,
    type,
    hideProgressBar,
    closeToast,
    transition: Transition,
    position,
    className,
    style,
    bodyClassName,
    bodyStyle,
    progressClassName,
    progressStyle,
    updateId,
    role,
    progress,
    rtl,
    toastId,
    deleteToast,
    isIn
  } = props;
  const defaultClassName = cx(
    `${Default.CSS_NAMESPACE}__toast`,
    `${Default.CSS_NAMESPACE}__toast--${type}`,
    {
      [`${Default.CSS_NAMESPACE}__toast--rtl`]: rtl
    }
  );
  const cssClasses = isFn(className)
    ? className({
        rtl,
        position,
        type,
        defaultClassName
      })
    : cx(defaultClassName, className);
  const isProgressControlled = !!progress;

  function renderCloseButton(closeButton: any) {
    if (!closeButton) return;

    const props = { closeToast, type };

    if (isFn(closeButton)) return closeButton(props);

    if (React.isValidElement(closeButton))
      return React.cloneElement(closeButton, props);
  }

  return (
    <Transition
      isIn={isIn}
      done={deleteToast}
      position={position}
      preventExitTransition={preventExitTransition}
      nodeRef={toastRef}
    >
      <div
        id={toastId as string}
        onClick={onClick}
        className={cssClasses}
        {...eventHandlers}
        style={style}
        ref={toastRef}
      >
        <div
          {...(isIn && { role: role })}
          className={
            isFn(bodyClassName)
              ? bodyClassName({ type })
              : cx(`${Default.CSS_NAMESPACE}__toast-body`, bodyClassName)
          }
          style={bodyStyle}
        >
          {children}
        </div>
        {renderCloseButton(closeButton)}
        {(autoClose || isProgressControlled) && (
          <ProgressBar
            {...(updateId && !isProgressControlled
              ? { key: `pb-${updateId}` }
              : {})}
            rtl={rtl}
            delay={autoClose as number}
            isRunning={isRunning}
            isIn={isIn}
            closeToast={closeToast}
            hide={hideProgressBar}
            type={type}
            style={progressStyle}
            className={progressClassName}
            controlledProgress={isProgressControlled}
            progress={progress}
          />
        )}
      </div>
    </Transition>
  );
};
