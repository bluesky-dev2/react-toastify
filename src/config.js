export default {
  POSITION: {
    TOP_LEFT: 'top-left',
    TOP_RIGHT: 'top-right',
    TOP_CENTER: 'top-center',
    BOTTOM_LEFT: 'bottom-left',
    BOTTOM_RIGHT: 'bottom-right',
    BOTTOM_CENTER: 'bottom-center'
  },
  TYPE: {
    INFO: 'info',
    SUCCESS: 'success',
    WARNING: 'warning',
    ERROR: 'error',
    DEFAULT: 'default'
  },
  ACTION: {
    SHOW: 'SHOW_TOAST',
    CLEAR: 'CLEAR_TOAST',
    MOUNTED: 'CONTAINER_MOUNTED'
  }
};
