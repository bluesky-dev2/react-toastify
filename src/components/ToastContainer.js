import React, {
  isValidElement,
  cloneElement,
  useState,
  useRef,
  useEffect
} from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import { TransitionGroup } from "react-transition-group";

import Toast from "./Toast";
import CloseButton from "./CloseButton";
import { Bounce } from "./Transitions";
import {
  POSITION,
  ACTION,
  RT_NAMESPACE,
  eventManager,
  falseOrDelay,
  isValidDelay,
  objectValues
} from "../utils";

export default function ToastContainer(props) {
  const [toast, setToast] = useState([]);
  const collection = useRef({});
  const containerRef = useRef(null);
  const toastKey = useRef(1);
  const ref = useRef();
  const r = useRef({
    containerId: props.containerId || null,
    isToastActive: false
  });

  function isToastActive(id) {
    return toast.indexOf(id) !== -1;
  }
  function removeToast(id){
    setToast(toast.filter(v => v !== id));
  }

  function dispatchChange(){
    eventManager.emit(ACTION.ON_CHANGE, toast.length, props.containerId);
    
  }

  function getAutoCloseDelay(toastAutoClose){
    return toastAutoClose === false || isValidDelay(toastAutoClose)
      ? toastAutoClose
      : props.autoClose;
  };

  //pure
  function canBeRendered(content) {
    return (
      isValidElement(content) ||
      typeof content === "string" ||
      typeof content === "number" ||
      typeof content === "function"
    );
  }

  function belongToContainer({ containerId }) {
    return containerId === props.containerId;
  }

  function buildToast(content, { delay, ...options }) {
    if (!canBeRendered(content)) {
      throw new Error(
        `The element you provided cannot be rendered. You provided an element of type ${typeof content}`
      );
    }

    const { toastId, updateId } = options;
    console.log({ toast });

    // Check for multi-container and also for duplicate toastId
    // Maybe it would be better to extract it
    if (
      (props.enableMultiContainer && !belongToContainer(options)) ||
      (isToastActive(toastId) && updateId == null)
    ) {
      return;
    }

    const closeToast = () => removeToast(toastId);
    const toastOptions = {
      id: toastId,
      // ⚠️ if no options.key, toastKey - 1 is assigned
      key: options.key || toastKey.current++,
      type: options.type,
      closeToast: closeToast,
      updateId: options.updateId,
      rtl: props.rtl,
      position: options.position || props.position,
      transition: options.transition || props.transition,
      className: parseClassName(options.className || props.toastClassName),
      bodyClassName: parseClassName(
        options.bodyClassName || props.bodyClassName
      ),
      onClick: options.onClick || props.onClick,
      closeButton: makeCloseButton(options.closeButton, toastId, options.type),
      pauseOnHover:
        typeof options.pauseOnHover === "boolean"
          ? options.pauseOnHover
          : props.pauseOnHover,
      pauseOnFocusLoss:
        typeof options.pauseOnFocusLoss === "boolean"
          ? options.pauseOnFocusLoss
          : props.pauseOnFocusLoss,
      draggable:
        typeof options.draggable === "boolean"
          ? options.draggable
          : props.draggable,
      draggablePercent:
        typeof options.draggablePercent === "number" &&
        !isNaN(options.draggablePercent)
          ? options.draggablePercent
          : props.draggablePercent,
      closeOnClick:
        typeof options.closeOnClick === "boolean"
          ? options.closeOnClick
          : props.closeOnClick,
      progressClassName: parseClassName(
        options.progressClassName || props.progressClassName
      ),
      progressStyle: props.progressStyle,
      autoClose: getAutoCloseDelay(options.autoClose),
      hideProgressBar:
        typeof options.hideProgressBar === "boolean"
          ? options.hideProgressBar
          : props.hideProgressBar,
      progress: parseFloat(options.progress),
      role: typeof options.role === "string" ? options.role : props.role
    };

    typeof options.onOpen === "function" &&
      (toastOptions.onOpen = options.onOpen);

    typeof options.onClose === "function" &&
      (toastOptions.onClose = options.onClose);

    // add closeToast function to react component only
    if (
      isValidElement(content) &&
      typeof content.type !== "string" &&
      typeof content.type !== "number"
    ) {
      content = cloneElement(content, {
        closeToast
      });
    } else if (typeof content === "function") {
      content = content({ closeToast });
    }

    if (isValidDelay(delay)) {
      setTimeout(() => {
        appendToast(toastOptions, content, options.staleToastId);
      }, delay);
    } else {
      appendToast(toastOptions, content, options.staleToastId);
    }
  }

  function appendToast(options, content, staleToastId) {
    const { id, updateId } = options;

    collection.current = {
      ...collection.current,
      [id]: {
        options,
        content,
        position: options.position
      }
    };
    setToast(
      (updateId ? [...toast] : [...toast, id]).filter(id => id !== staleToastId)
    );
  }

  function renderToast() {
    const toastToRender = {};
    const { className, style, newestOnTop } = props;
    const currentCol = collection.current;
    const col2 = newestOnTop
      ? Object.keys(currentCol).reverse()
      : Object.keys(currentCol);

    // group toast by position
    col2.forEach(toastId => {
      const { position, options, content } = currentCol[toastId];

      toastToRender[position] || (toastToRender[position] = []);

      if (toast.indexOf(options.id) !== -1) {
        toastToRender[position].push(
          <Toast {...options} key={`toast-${options.key}`}>
            {content}
          </Toast>
        );
      } else {
        toastToRender[position].push(null);
        delete collection.current[toastId];
      }
    });

    return Object.keys(toastToRender).map(position => {
      const disablePointer =
        toastToRender[position].length === 1 &&
        toastToRender[position][0] === null;
      const propsp = {
        className: cx(
          `${RT_NAMESPACE}__toast-container`,
          `${RT_NAMESPACE}__toast-container--${position}`,
          { [`${RT_NAMESPACE}__toast-container--rtl`]: props.rtl },
          parseClassName(className)
        ),
        style: disablePointer
          ? { ...style, pointerEvents: "none" }
          : { ...style }
      };

      return (
        <TransitionGroup {...propsp} key={`container-${position}`}>
          {toastToRender[position]}
        </TransitionGroup>
      );
    });
  }

  function clear() {
    setToast([]);
  }

  function parseClassName(prop) {
    if (typeof prop === "string") {
      return prop;
    } else if (
      prop !== null &&
      typeof prop === "object" &&
      "toString" in prop
    ) {
      return prop.toString();
    }

    return null;
  }

  function makeCloseButton(toastClose, toastId, type) {
    let closeButton = props.closeButton;

    if (isValidElement(toastClose) || toastClose === false) {
      closeButton = toastClose;
    } else if (toastClose === true) {
      closeButton =
        props.closeButton && typeof props.closeButton !== "boolean" ? (
          props.closeButton
        ) : (
          <CloseButton />
        );
    }

    return closeButton === false
      ? false
      : cloneElement(closeButton, {
          closeToast: () => removeToast(toastId),
          type: type
        });
  }
  useEffect(() => {
    console.log("UPDATE BUIDL");

    ref.current = buildToast;
  }, [props, toast]);

  useEffect(() => {
    console.log("ICI", toast);

    dispatchChange();
  }, [toast]);

  useEffect(() => {
    console.log("effect");

    r.current.isToastActive = isToastActive;
    eventManager
      .cancelEmit(ACTION.WILL_UNMOUNT)
      .on(ACTION.SHOW, (content, options) =>
        containerRef.current ? ref.current(content, options) : null
      )
      .on(ACTION.CLEAR, id =>
        !containerRef.current ? null : id == null ? clear() : removeToast(id)
      )
      .emit(ACTION.DID_MOUNT, r.current);

    return () => eventManager.emit(ACTION.WILL_UNMOUNT, r.current);
  }, []);

  return (
    <div ref={containerRef} className={`${RT_NAMESPACE}`}>
      {renderToast()}
    </div>
  );
}

ToastContainer.propTypes = {
  /**
   * Set toast position
   */
  position: PropTypes.oneOf(objectValues(POSITION)),

  /**
   * Disable or set autoClose delay
   */
  autoClose: falseOrDelay,

  /**
   * Disable or set a custom react element for the close button
   */
  closeButton: PropTypes.oneOfType([PropTypes.node, PropTypes.bool]),

  /**
   * Hide or not progress bar when autoClose is enabled
   */
  hideProgressBar: PropTypes.bool,

  /**
   * Pause toast duration on hover
   */
  pauseOnHover: PropTypes.bool,

  /**
   * Dismiss toast on click
   */
  closeOnClick: PropTypes.bool,

  /**
   * Newest on top
   */
  newestOnTop: PropTypes.bool,

  /**
   * An optional className
   */
  className: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),

  /**
   * An optional style
   */
  style: PropTypes.object,

  /**
   * An optional className for the toast
   */
  toastClassName: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),

  /**
   * An optional className for the toast body
   */
  bodyClassName: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),

  /**
   * An optional className for the toast progress bar
   */
  progressClassName: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),

  /**
   * An optional style for the toast progress bar
   */
  progressStyle: PropTypes.object,

  /**
   * Define enter and exit transition using react-transition-group
   */
  transition: PropTypes.func,

  /**
   * Support rtl display
   */
  rtl: PropTypes.bool,

  /**
   * Allow toast to be draggable
   */
  draggable: PropTypes.bool,

  /**
   * The percentage of the toast's width it takes for a drag to dismiss a toast
   */
  draggablePercent: PropTypes.number,

  /**
   * Pause the toast on focus loss
   */
  pauseOnFocusLoss: PropTypes.bool,

  /**
   * Show the toast only if it includes containerId and it's the same as containerId
   */
  enableMultiContainer: PropTypes.bool,

  /**
   * Set id to handle multiple container
   */
  containerId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

  /**
   * Set role attribute for the toast body
   */
  role: PropTypes.string,

  /**
   * Fired when clicking inside toaster
   */
  onClick: PropTypes.func
};

ToastContainer.defaultProps = {
  position: POSITION.TOP_RIGHT,
  transition: Bounce,
  rtl: false,
  autoClose: 5000,
  hideProgressBar: false,
  closeButton: <CloseButton />,
  pauseOnHover: true,
  pauseOnFocusLoss: true,
  closeOnClick: true,
  newestOnTop: false,
  draggable: true,
  draggablePercent: 80,
  className: null,
  style: null,
  toastClassName: null,
  bodyClassName: null,
  progressClassName: null,
  progressStyle: null,
  role: "alert"
};
