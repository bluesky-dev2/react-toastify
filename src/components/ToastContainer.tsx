import * as React from 'react';
import cx from 'clsx';

import { Toast } from './Toast';
import { CloseButton } from './CloseButton';
import { Bounce } from './Transitions';
import { POSITION, Direction, Default, parseClassName, isFn } from '../utils';
import { useToastContainer } from '../hooks';
import { ToastContainerProps, ToastPosition } from '../types';

export const ToastContainer: React.FC<ToastContainerProps> = props => {
  const { getToastToRender, containerRef, isToastActive } = useToastContainer(
    props
  );
  const { className, style, rtl, containerId } = props;

  function getClassName(position: ToastPosition) {
    const defaultClassName = cx(
      `${Default.CSS_NAMESPACE}__toast-container`,
      `${Default.CSS_NAMESPACE}__toast-container--${position}`,
      { [`${Default.CSS_NAMESPACE}__toast-container--rtl`]: rtl }
    );
    return isFn(className)
      ? className({
          position,
          rtl,
          defaultClassName
        })
      : cx(defaultClassName, parseClassName(className));
  }

  return (
    <div
      ref={containerRef}
      className={Default.CSS_NAMESPACE as string}
      id={containerId as string}
    >
      {getToastToRender((position, toastList) => {
        const containerStyle: React.CSSProperties =
          toastList.length === 0
            ? { ...style, pointerEvents: 'none' }
            : { ...style };

        return (
          <div
            className={getClassName(position)}
            style={containerStyle}
            key={`container-${position}`}
          >
            {toastList.map(({ content, props: toastProps }) => {
              return (
                <Toast
                  {...toastProps}
                  isIn={isToastActive(toastProps.toastId)}
                  key={`toast-${toastProps.key}`}
                  closeButton={
                    toastProps.closeButton === true
                      ? CloseButton
                      : toastProps.closeButton
                  }
                >
                  {content}
                </Toast>
              );
            })}
          </div>
        );
      })}
    </div>
  );
};

ToastContainer.defaultProps = {
  position: POSITION.TOP_RIGHT as ToastPosition,
  transition: Bounce,
  rtl: false,
  autoClose: 5000,
  hideProgressBar: false,
  closeButton: CloseButton,
  pauseOnHover: true,
  pauseOnFocusLoss: true,
  closeOnClick: true,
  newestOnTop: false,
  draggable: true,
  draggablePercent: Default.DRAGGABLE_PERCENT as number,
  draggableDirection: Direction.X,
  role: 'alert',
  theme: 'light'
};
