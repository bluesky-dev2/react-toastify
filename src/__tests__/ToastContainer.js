/* eslint-env jest */
import React from 'react';
import { mount } from 'enzyme';

import ToastContainer from './../ToastContainer';
import Toast from './../Toast';
import toaster from './../toaster';

import config from './../config';
import EventManager from './../util/EventManager';

function hasProp(obj, key) {
  return Object.prototype.hasOwnProperty.call(obj, key);
}

jest.useFakeTimers();

describe('ToastContainer', () => {
  it('Should merge className and style', () => {
    const component = mount(<ToastContainer className="plop" style={{ background: 'red' }} />);
    const expectedStyle = { background: 'red' };

    toaster('coucou');
    jest.runAllTimers();

    expect(component.find('.toastify').hasClass('plop')).toBe(true);
    expect(component.find('.toastify').prop('style')).toMatchObject(expectedStyle);
  });

  it('Should bind event when mounted and unbind them when unmounted', () => {
    const component = mount(<ToastContainer />);

    expect(EventManager.eventList.has(config.ACTION.SHOW)).toBeTruthy();
    expect(EventManager.eventList.has(config.ACTION.CLEAR)).toBeTruthy();

    component.unmount();
    expect(EventManager.eventList.has(config.ACTION.SHOW)).toBeFalsy();
    expect(EventManager.eventList.has(config.ACTION.CLEAR)).toBeFalsy();
  });

  it('Should set style pointer events to none when there is no toast to render', () => {
    const component = mount(<ToastContainer />);

    toaster('coucou');
    jest.runAllTimers();
    toaster.dismiss();
    jest.runAllTimers();

    expect(component.find('.toastify').prop('style').pointerEvents).toEqual('none');
  });

  it(`Should always pass down to Toast the props: 
    -autoClose
    -closeButton
    -children
    -position
    -pauseOnHover
    -closeToast`, () => {
    const component = mount(<ToastContainer />);
    // Create a toast
    toaster('coucou');
    jest.runAllTimers();

    const props = component.children().find(Toast).props();

    [
      'autoClose',
      'closeButton',
      'children',
      'position',
      'closeToast',
      'pauseOnHover'
    ].forEach(key => expect(hasProp(props, key)).toBeTruthy());
  });
});
