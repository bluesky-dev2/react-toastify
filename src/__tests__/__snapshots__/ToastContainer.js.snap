// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToastContainer Should allow glamor rule as className 1`] = `
.glamor-0,
[data-glamor-0] {
  background: red;
}

<ToastContainer
  autoClose={5000}
  bodyClassName={null}
  className={
    Object {
      "data-glamor-0": "",
    }
  }
  closeButton={
    <DefaultCloseButton
      ariaLabel="close"
    />
  }
  closeOnClick={true}
  hideProgressBar={false}
  newestOnTop={false}
  pauseOnHover={true}
  pauseOnVisibilityChange={true}
  position="top-right"
  progressClassName={null}
  rtl={false}
  style={null}
  toastClassName={null}
  transition={[Function]}
>
  <div />
</ToastContainer>
`;
